Modules/gcmodule.c

```diff
@@ -64,74 +114,63 @@ _PyGC_Initialize(struct _gc_runtime_state *state)
+Between collections, gc_prev is used for doubly linked list.
 
-GC_REACHABLE
-    The object lives in some generation list, and its tp_traverse is safe to
-    call.  An object transitions to GC_REACHABLE when PyObject_GC_Track
-    is called.
+Lowest three bits of gc_prev are used for flags.
+MASK_COLLECTING and MASK_TENTATIVELY_UNREACHABLE are used only while
+collecting and cleared before GC ends or _PyObject_GC_UNTRACK() is called.
 
-During a collection, gc_refs can temporarily take on other states:
+During a collection, gc_prev is temporary used for gc_refs, and the gc list
+is singly linked until gc_prev is restored.
 
->= 0
+gc_refs
     At the start of a collection, update_refs() copies the true refcount
     to gc_refs, for each object in the generation being collected.
     subtract_refs() then adjusts gc_refs so that it equals the number of
     times an object is referenced directly from outside the generation
     being collected.
-    gc_refs remains >= 0 throughout these steps.
 
-GC_TENTATIVELY_UNREACHABLE
+MASK_COLLECTING
+    Objects in generation being collected are marked MASK_COLLECTING in
+    update_refs().
+
+MASK_TENTATIVELY_UNREACHABLE
     move_unreachable() then moves objects not reachable (whether directly or
-    indirectly) from outside the generation into an "unreachable" set.
-    Objects that are found to be reachable have gc_refs set to GC_REACHABLE
-    again.  Objects that are found to be unreachable have gc_refs set to
-    GC_TENTATIVELY_UNREACHABLE.  It's "tentatively" because the pass doing
-    this can't be sure until it ends, and GC_TENTATIVELY_UNREACHABLE may
-    transition back to GC_REACHABLE.
-
-    Only objects with GC_TENTATIVELY_UNREACHABLE still set are candidates
-    for collection.  If it's decided not to collect such an object (e.g.,
-    it has a __del__ method), its gc_refs is restored to GC_REACHABLE again.
+    indirectly) from outside the generation into an "unreachable" set and
+    set MASK_TENTATIVELY_UNREACHABLE flag.
+
+    Objects that are found to be reachable have gc_refs set to 1.
+    When MASK_TENTATIVELY_UNREACHABLE flag is set for the reachable object,
+    the flag is unset and the object is moved back to "reachable" set.
+
+    Only objects with MASK_TENTATIVELY_UNREACHABLE still set are candidates
+    for collection.
 ----------------------------------------------------------------------------
 */
-#define GC_UNTRACKED                    _PyGC_REFS_UNTRACKED
-#define GC_REACHABLE                    _PyGC_REFS_REACHABLE
-#define GC_TENTATIVELY_UNREACHABLE      _PyGC_REFS_TENTATIVELY_UNREACHABLE
-
-#define IS_TRACKED(o) (_PyGC_REFS(o) != GC_UNTRACKED)
-#define IS_REACHABLE(o) (_PyGC_REFS(o) == GC_REACHABLE)
-#define IS_TENTATIVELY_UNREACHABLE(o) ( \
-    _PyGC_REFS(o) == GC_TENTATIVELY_UNREACHABLE)
 
 /*** list functions ***/
 
 static void
 gc_list_init(PyGC_Head *list)
 {
-    list->gc.gc_prev = list;
+    list->gc.gc_prev = (uintptr_t)list;
```

Comment:
I would prefer to use a macro here since the field is special. _PyGCHead_SET_PREV_FLAGS(list, 0)?

Ditto below.

Reply:
This code doesn't mean "clear flags".  List header node (gc_list) doesn't have any flags.
So gc_prev is really pure pointer without any flags here.
I'll add one line comment here too.